demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+// Copyright (C) 2019 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "cpu_monitor.h"
+#ifdef WIN32
+#include <tchar.h>
+
+namespace {
+std::size_t getNCores() {
+    SYSTEM_INFO sysinfo;
+    GetSystemInfo(&sysinfo);
+    return sysinfo.dwNumberOfProcessors;
+}
+}
+
+CpuMonitor::CpuMonitor() : nCores{getNCores()}, lastEnabled{false}, historyEnabled{false}, samplesNumber{0} {};
+
+void CpuMonitor::openQuery() {
+    PDH_STATUS status;
+
+    status = PdhOpenQuery(NULL, NULL, &query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhOpenQuery() failed");
+    }
+
+    for (std::size_t i = 0; i < nCores; ++i)
+    {
+        TCHAR szCounterName[MAX_PATH];
+        _stprintf_s(szCounterName, sizeof(szCounterName), TEXT("\\Processor(%u)\\%% Processor Time"), i);
+
+        PDH_HCOUNTER counter;
+
+        status = PdhAddCounter(query, szCounterName, 0, &counter);
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhAddCounter() failed");
```

Comment:
updated

Reply:
You have added a wrapper class, but you haven't really solved the problem: it's still possible for the query to remain allocated if `CpuMonitor::openQuery` fails.

Reply:
Will, kind of updated.
I think it is better to close query in case of errors manually, because now I open/close extra query to override previous extra query :)

Reply:
Your current implementation is not correct, because if you call `closeQuery` and then destroy the object, it will attempt to close the handle again.

I suggest you make use of `unique_ptr` for this:

```c++
class CpuMonitor {
    // [...]
    std::unique_ptr<QueryWrapper> query;
};

void CpuMonitor::openQuery() {
    std::unique_ptr<QueryWrapper> newQuery(new QueryWrapper);
    // [... configure counters ...]
    query = std::move(newQuery);
}

void CpuMonitor::closeQuery() {
    coreTimeCounters.clear();
    query.reset();
}
```

As a bonus, this will allow you to move the definition of `QueryWrapper` out of the header file.

Reply:
updated.
Moved `QueryWrapper` to a separate files cos both monitors now use it.